### ISSUE TITLE
The automatic behavior of `python.analysis.exclude` is ambiguous when I specify excluded paths.

### ISSUE THREAD

**Issue by Y-Haneji** [microsoft/pylance-release#5970]
The description of `python.analysis.exclude` says
> If no exclude paths are specified, Pylance automatically excludes the following: **/node_modules, **/__pycache__, .git and any virtual environment directories.

Does Pylance still automatically exclude these directories when I specify some excluded paths?
Or, do I have to explicitly specify them to exclude?

If yes, what does the "virtual environment directories" concretely indicate?

**Comment by rchiodo** [comment]
Thanks for the issue. Yeah I agree, it's confusing. We should fix the docs. 

I can explain it better here though. 

> Does Pylance still automatically exclude these directories when I specify some excluded paths?
Or, do I have to explicitly specify them to exclude?

TLDR; You have to explicitly specify them. 

The logic for all of this is actually open source if you want to look at it:

First if the exclude list is empty, these items ('**/node_modules', '**/__pycache__', '**/.*') are added to the exclude list:
https://github.com/microsoft/pyright/blob/8fa827608dfd4693233ad41b0081a0f1f39ca486/packages/pyright-internal/src/analyzer/service.ts#L599

You can see the logic where it adds that default list to the exclusions a couple of lines down:
https://github.com/microsoft/pyright/blob/8fa827608dfd4693233ad41b0081a0f1f39ca486/packages/pyright-internal/src/analyzer/service.ts#L690

> If yes, what does the "virtual environment directories" concretely indicate?

The virtual environment logic is a bit more complicated, but it also only applies when the exclude list is empty. If the exclude list is empty, the `autoExcludeVenv` flag is set here:
https://github.com/microsoft/pyright/blob/8fa827608dfd4693233ad41b0081a0f1f39ca486/packages/pyright-internal/src/analyzer/service.ts#L696

This is later used when we look for files in the workspace to exclude virtual environment type directories. The virtual environment type directories are discovered by having specific files in them. That list is here:
https://github.com/microsoft/pyright/blob/8fa827608dfd4693233ad41b0081a0f1f39ca486/packages/pyright-internal/src/analyzer/service.ts#L1188

So when the exclusion list is empty, we will also exclude any root folder that has a:
- bin/activate sub folder
- Scripts/activate sub folder
- pyvenv.cfg file
- conda-meta sub folder

So if wanted to replicate this behavior, you'd have to explicitly exclude your venv.

**Comment by Y-Haneji** [comment]
I see. It clarified the automatic behavior well.

It's better to set the default excluded items (** /node_modules, ** /pycache, .git and any virtual environment directories) as the default setting values, like the `files.exclude` setting.

Anyway, thank you for your kind answer!

**Comment by HAL-42** [comment]
It seems that both `search.exclude` and `files.exclude` preserve the default exclusion behavior; `python.analysis.exclude` should maintain consistent behavior.

**Comment by jaens** [comment]
The current behavior is a "nice" footgun, after adding this setting to the workspace, Pylance immediately started analyzing the _entire_ Python installation in `.venv` since the default excludes stopped working...

**Comment by heejaechang** [comment]
This issue has been fixed in prerelease version 2026.3.100, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20263100-24-August-2026-PreRelease)
